### ISSUE TITLE
`MessageEncrypter` should have a non-allocating interface for `encrypt`

### ISSUE THREAD

**Issue by briansmith** [rustls/rustls#670]
Currently we have only:
```
    fn encrypt(&self, m: BorrowMessage, seq: u64) -> Result<Message, Error>;
```

We should have a variant that works like this:
```
    fn encrypt<P>(&self, m: Message, seq: u64) -> Result<Message, Error>;
```
This new variant would be able to avoid copying the message payload into a new `Vec` because it would be able to reuse the `Vec` in `m` and use the in-place *ring* API for encryption that as designed for this.

**Comment by djc** [comment]
I've been looking at this, too. I think the issue here is the insertion of nonces. We can probably do it in a smarter way, though, and I plan to work on that.

**Comment by briansmith** [comment]
> I think the issue here is the insertion of nonces. We can probably do it in a smarter way, though, and I plan to work on that.

First, the explicit part of the nonce is only inserted for TLS 1.2, not TLS 1.3. So, we need not worry to much about it in terms of performance since the TLS 1.3 path is the one that we should be focusing on optimizing. Regardless, I would expect a new API that takes Vec without doing anything smarter would still be at least as efficient even for the annoying TLS 1.2 case.

**Comment by ctz** [comment]
I'm not sure this would be measurable in benchmarks, but it'd be very interesting to find out. Ultimately the most important benchmark involves taking non-owned bytes and encrypting them, which is why the interface is how it currently is. But perhaps this would affect the TLS1.3 full handshake measurements (as that involves a large encryption of owned bytes -- the server and client certificate chains).

**Comment by alcore** [comment]
+1 on @briansmith's experience.

 > (...) Ultimately the most important benchmark involves taking non-owned bytes and encrypting them (...)
 
As it is currently, rustls **_owns_ the data _anyways_** due to at least 2 (IIRC) unnecessary allocating copies of the appdata passed in, with many more unnecessary copies for its own protocol messages during the handshake. The issue is further exacerbated by how scatter/gather IO is used in `VecChunkBuf`, where the current design entirely defeats the purpose of using vectored IO in the first place. As it is, it only adds additional unnecessary overhead and commits like https://github.com/ctz/rustls/commit/62d15df98e6830fee74d6155f28f14dfae1b0187 compound on that, where now it's stack allocating 1KB of empty slices for _each_ single write.

Since rustls _is buffering anyways_, the IMHO sane approach would be either to utilize this fact, or work towards a proper zero-copying implementation with vectored writes. 

Instead of `VecChunkBuf` just write all messages to a contiguous byte buffer. Encode them straight into that buffer, encrypt them in that buffer, append the MACs/tags to that buffer - so forth. There's really only one non-allocating copy needed for every write (barring the occasional grow should the buffer need to), even with the current overall design, regardless of whether appdata needs to be chunked or encrypted.

We made a fork-turned-full-rewrite about half a year ago starting with precisely this change.

In any case - don't needlessly invoke scatter/gather IO. With a scatter/gather approach we'd want a mutable slice from the app, as @briansmith suggested, and simply prepend the frame header (in the case of v1.2 GCM also the nonce) as an IoSlice to the encrypted text as an IoSlice, followed by the MAC/tag as the third IoSlice. In this case I'd further expect all io::Write calls to be non-buffering and invoke a `write_vectored` call directly - buffering my data is a side effect I don't expect from a TLS library. 

However, ultimately some sort of buffer is required anyways (be it to support replaying the transcript for client auth or to tackle incomplete writes/reads), so I'd suggest dropping scatter/gather, at least for now, and sticking to the buffer approach - but actually _embrace_ it instead of trying to work around it in ways that do nothing but introduce additional overhead and complexity. Coupling chunking, framing, and buffering in a `Writer` type that has access to the cipher solved this in our case. It cut down on both code complexity and allocs.

@briansmith What I think would be helpful (and would love to see, so we can use that in our fork as well) is a variant of `seal_in_place_separate_tag` that is not in place but takes a immutable `src` and a mutable `dst`. In-place encryption works great if we could fully leverage vectored I/O, but in a design which buffers anyways, `seal_in_place_separate_tag` is currently one extra copy of the app data as it gets written to said buffer, considering the input comes (likely) from io::Write and is immutable. Since we're writing the TLS framing to the buffer and need to append the message (and TLS1.2 GCM nonce) anyways, writing to said buffer as ring goes would cut that copy down.

> (...) But perhaps this would affect the TLS1.3 full handshake measurements (as that involves a large encryption of owned bytes -- the server and client certificate chains).(...)

It applies to every message encrypted during the session's lifetime, and I'd argue most applications will encrypt considerably more data than the lib will during the handshake - and, IMHO, _that_ should be the benchmark to refer to.

In particular, we've got long-lived (minutes up to hours) connections open which is why - and this is also related to https://github.com/ctz/rustls/issues/665 as memory became a concern as well - we swap out the entire session from a lib-owned "HandshakeSession" into an "AuthenticatedSession" once the handshake is done, dropping all handshake state in the process and moving to ApplicationData (instead of Handshake) optimized codepaths instead. The opened up _a lot_ of optimization opportunities.

Post-handshake connections retain a minimal protocol state handler on top - for the odd case when we do get a Handshake frame - which in the case of TLS1.2 is effectively a no-op for both reads and writes (= no renegotiation) while for TLS1.3 it retains only the transcript's final handshake digest as state (so it can handle potential key updates). All other data is dropped, including the config but except for a cache accessor (again, TLS 1.3 tickets received post-handshake).

All of it does matter _a lot_ when you start looking into hundreds, let alone thousands, of open connections. I don't have the numbers on hand, but IIRC it was about a 60-80% cut in memory usage per conn (where memory usage per conn is in the kilobytes in rustls' case, _plus several_ I/O buffers that can't be reused and managed by the app). There's a cost in code duplication to be paid for this approach, but it was well worth it in our case. While we did go unsafe ultimately, and on the other part rewrote the entire state handling flow, IIRC from initial benchmarks the biggest gains were actually from very low hanging fruit and bringing IO much closer to `ring`'s ciphers was part of it.

Memory wise - outside of the split into handshake and post-handshake states - we've also had nice gains from moving things like client/server extensions, signature algs, suites et al into tiny bitsets instead of vectors, saving on both the allocs and overall memory use, but also cutting down all the O(n) (or in some cases even quadratic) comparisons during negotiation, since those can be handled by simple bitmask intersections. Additionally, a sufficiently considerable amount of the data sent and/or expected is actually constant and there's needless overhead in both encoding and decoding such constant data (into an allocating Vec at that) during each handshake. We simplified that mostly by coupling decoding with validation on concrete types, instead of a generic approach to enc/dec. 

Similarly, encoding what effectively is `pub const CCS: [u8; 6] = [frame::Type::ChangeCipherSpec as u8, 3, 3, 0, 1, 1];` and allocating that as a separate Vec each time a CCS is sent out felt incredibly wasteful - but wasn't possible to do otherwise without rewriting rustls' internals. I'm aware those few dozen nanoseconds are swamped by both encryption and IO, but the thing is: there's a lot of those missed opportunities in the current design and those ns quickly turn into μs.

**Comment by briansmith** [comment]
@alcore Awesome thoughts! I hope we can find some way to work together on this because those thoughts are very similar to the ones I have had recently.

> cutting down all the O(n) (or in some cases even quadratic) comparisons during negotiation,

I think this did get a little better recently as part of the refactoring of the handshake that's being done. It would be great to see an example of an improvement of the sort you're talking about, ideally as a PR, or otherwise as a separate issue.

**Comment by ctz** [comment]
> As it is currently, rustls **_owns_ the data _anyways_** due to at least 2 (IIRC) unnecessary allocating copies of the appdata passed in

Could you point out where this is happening please?

**Comment by briansmith** [comment]
> @briansmith What I think would be helpful (and would love to see, so we can use that in our fork as well) is a variant of `seal_in_place_separate_tag` that is not in place but takes a immutable `src` and a mutable `dst`.

Let's discuss that within the *ring* project. I know how I want to add such an API without adding a bunch of code bloat. Regardless, it wouldn't be hard to add the API and then optimize the waste away later.

In the context of Rustls, I think saying we need that kind of interface may be putting the cart in front of the horse. I think we'd need to sort out an iovec-like public API for applications to write (and read) through Rustls, with a clear plan for how we can avoid the allocations and copies that happen internally.

As a baby step to that, I do think we should consider redoing `Message` and `OpaqueMessage` and `BorrowMessage` something like this:

```
struct Message<B: AsMut<[u8]>> {
    header: [u8; HEADER_LENGH + TLS12_EXPLICIT_NONCE_MAX_LEN],
    payload: B,
}
```

In other words, we should consider making each `Message` an iovec so we don't have to do memcpys to deal with headers and (in TLS 1.2) explicit nonce fragments.

**Comment by alcore** [comment]
> > As it is currently, rustls **_owns_ the data _anyways_** due to at least 2 (IIRC) unnecessary allocating copies of the appdata passed in
> 
> Could you point out where this is happening please?

My bad, memory did not serve me well - besides the cipher alloc, I misremembered `send_plain` from `SessionCommon`, but that doesn't apply post-handshake. We entirely dropped the ability to R/W before the handshake completes, maybe thus my confusion.

There's non-guaranteed allocating copies remaining, however. E.g.,
-  _if/when_ chunking (https://github.com/ctz/rustls/blob/main/rustls/src/msgs/fragmenter.rs#L43). 
-  VecDeque allocation(s) (https://github.com/ctz/rustls/blob/main/rustls/src/conn.rs#L790 and others) that happen regardless of whether the message gets chunked or not. It's small and not a data copy, but still feels unnecessary.

I still need to catch up on the current refactoring effort, but IIRC each handshake message was being allocated more than once too (not counting the potential encryption alloc). This was due to encoding multiple times. Once in the transcript (https://github.com/ctz/rustls/blob/main/rustls/src/hash_hs.rs#L79), another one due to the calls to `.into_opaque()` (https://github.com/ctz/rustls/blob/main/rustls/src/msgs/message.rs#L199) in several code paths. It's essentially 2*2 allocating copies, due to https://github.com/ctz/rustls/blob/main/rustls/src/msgs/handshake.rs#L2172 allocating separately. There were also unexpected (to us) side effects, like a seemingly trivial .len() call actually encoding the entire message (allocating) just to return its length (https://github.com/ctz/rustls/blob/main/rustls/src/msgs/handshake.rs#L2190).

This one was low hanging fruit. We simply exposed a `transcribe` method on the Session to the state handlers. It'd run the encoding once and pass the resulting slice to the transcript, then to the (chunking) buffer. Much later on we could get rid of the transcript's buffer entirely too, but that was past the rewrite.

Reads had similar issues, allocating while decoding, despite owning a full frame buffer in the Reader/Deframer, but IIRC we had to do an unsafe lifetime erasure to get the borrowed frames into the state handlers without Big-Daddy-Borrow-Checker stepping in -- at least until the rewrite was mostly done and state handlers no longer needed access to the entirety of the Session's state (which included the reader with its buffer, and thus aliasing was happening).

I apologize if some of it is out of date already.

> As a baby step to that, I do think we should consider redoing Message and OpaqueMessage and BorrowMessage something like this (...)

I'd +1 that proposal, but I believe it'd be overlooking the fact that TLS1.3 ciphers need to mask the header and append the actual frame type _after_ the payload **and encrypt it** along with the payload. `AsMut` therefore won't cut it, because we'd need `Extend` of some sort to pass that to `ring` as a contiguous message.

Further, even disregarding the above, ciphers would need the ability to either:
- write to a `&mut Vec<IoSlice>` or `&mut [3; IoSlice]`
- return a `[3; IoSlice]`.

3 slices: prefix (header sliced with or without nonce), payload and tag (disregarding the issue with TLS1.3 masking for now).

 `ChunkVecBuffer` can't currently queue such messages since it's a `VecDeque<Vec<u8>>`, where we'd at least need a `Vec<IoSlice>`. So we either write that out directly without queueing, on each io::Write (potentially breaking user assumptions about rustls' internal buffering behavior) or we keep a separate Vec<IoSlice> allocated for AppData messages from userland (we currently still need that `VecDeque<Vec<u8>>` for rustls' own messages -- and refactoring encoding is a pain story of its own).
 
Further, `AsMut` isn't compatible with io::Write so this would need to be a separate API. At the same time, the non-mutable io::Write would still need to be supported, and to support that, at least without unsafe, we need a means of taking ownership of that data, which by default means we need a copy, which in turn brings us back to a single buffer being the simplest approach. Supporting both APIs without feature gating and fairly complex conditional compilation would mean that users who exclusively use the vectored API still pay the price of retaining the buffer rustls needs anyways (at the very least during the handshake - and currently until a `drop()` of the Connection) while users who exclusively use the buffering API end up on a codepath that aims to force them through a vectored write, adding overhead unless we take on the maintenance burden of keeping those separate all the way. Sure, that overhead would be negligible relative to encryption/IO, but it's still there and it's paying for something that I don't use.

With all of the above in mind, I still suggest simply embracing run of the mill read and write buffers to get this rolling. I/we couldn't think of a way to implement proper zero-copy scatter/gather without ring accepting a slice of mutable byte slices as contiguous memory for encryption (due to the issues mentioned earlier). At the same time I'm rather confident @briansmith won't want to support that because I assume it'd complicate ring's internals by a factor of `oh hello no`. _But_ I have a tendency to miss obvious things.

My suggestion is to also consider exploring the option of breaking the current API entirely and having rustls take full control over the `io::Write + io::Read` (socket, presumably) it gets from the app until the handshake actually yields a `Result<Connection, Error>` of sorts -- getting rid of as much buffering and special-cased execution flows as possible (and opening up optimizations for separate handshake and traffic states through the type system).

<details>
  <summary>Our simplified approach</summary>
WRT to this particular issue, we went from a `struct Frame(Vec<u8>);` wrapper with an interface along the lines of...

```rust
impl Frame {
    pub fn typ(&self) -> frame::Type
    pub fn header(&self) -> frame::Header
    pub fn payload(&self) -> &[u8]
    pub fn payload_mut(&mut self) -> &mut [u8]
    pub fn len(&self) -> usize
   (...)
}
```

... The ciphers were responsible for writing that to the final buffer in their own order (vide TLS1.2 GCM) and passing what's needed to `ring`. This wasn't aimed at perf (we still needed a copying alloc due to encoding anyways), but was a step towards refactoring the entirety of the encoding flow and separating state. Once we wrapped that up, we got rid of that alloc too, encoding messages straight into a Session-owned buffer and only passed...

```rust
struct EncryptionContext<'a> {
    buf: &'a mut Vec<u8>,
    idx: usize
}
```

... to the ciphers. That struct has basically the same API for convenience and simply interprets the bytes within the `idx ... buf.len()` range as a `Frame`. There's some code duplication in the ciphers since they need to each understand the encoding and handle it it in their own way, but ultimately they remained trivial. TLS1.2 GCM in our case is feature gated (as are all suites), but when it's used, we zero-prepend the payload before copying it in and passing it to the cipher, which at least allows us to avoid a further memmove of the payload.

Sure, vectored I/O would be lovely, but those were the fruits we could actually reach.
</details>

<details>
  <summary>We should clearly document the IO behavior of rustls</summary>

Depending on the direction, if buffering ApplicationData stays roughly as it is, I think it'd be beneficial to document the behavior more clearly. For example, make it clear - especially to new users - that explicitly using `write_all`, `write_vectored` and/or the unstable `write_all_vectored` is effectively just added overhead since calls to io::Write fully buffer the inputs. The current docs phrase this as:

> Note that client.write() buffers data you send _if the TLS session is not yet established_: this is useful for writing (say) a HTTP request, but don’t write huge amounts of data.`

So we had a gotcha moment upon digging in. The fact that buffering actually takes place _always_ can be a _useful_ property to someone conscious of I/O and syscall overheads but one can't utilize this usefulness unless one knows about it. You can also shoot yourself in the foot unknowingly by assuming you're io::Writing data that you buffer yourself (as common in a HTTP context) where in reality you're filling up memory that you're not flushing to the actual socket.

</details>

<details>
  <summary>Offtopic: Regarding contributing actual code</summary>

> It would be great to see an example of an improvement of the sort you're talking about, ideally as a PR, or otherwise as a separate issue.

Yep, apologies for barging in like this with such lengthy comments, nitpicking without code. None of it made it upstream because while we started small, this turning into a rewrite meant we broke the entire API and I didn't feel I could push such changes - dev on rustls wasn't as lively half a year ago as it currently seems to be.

I'll start digging in and see if there's something I can port back over. We have differing and considerably simpler assumptions (e.g., no QUIC - also meaning no out-of-order frames, no early data, and essentially only Connections are part of the public API), but some should be possible. 

But, some won't be. 

For instance, we have ciphersuites as a u16 bitmask - this makes the related logic a pleasure to work with **but** it also means that the order of preference is dictated by the order we encode them in. This in turn looks like this:

```rust
impl SuiteSet {
    (...)
    pub const fn encoded_len(&self) -> u16 {
        self.0.count_ones() as u16 * Suite::SIZE as u16
    }

    pub unsafe fn encode(self, mut dst: *mut u8) -> *mut u8 {
        // In no code path should we be serializing an empty suite set.
        debug_assert!(self.0 != 0);

        let mut set = self.0;
        while set != 0 {
            let m = set & set.wrapping_neg();
            // SAFETY: Suite has a fixed repr(u16) and we are guarding against
            // a zero value (Unknown) in the loop condition.
            let suite: Suite = std::mem::transmute(m);
            dst = suite.encode(dst);
            set ^= m;
        }

        dst
    }
    (...)
}
```

... so, not even taking the unsafe and our frequent use of several `const`-related nightly features into account, I can't straight up port any of it over without making sure I don't accidentally attempt to port over my own assumptions along with it.

We use bitmasks extensively but I'll need to explore all of them in separate issues/PRs, since they come with several caveats which might make them unfeasible to a more generic purpose (and open) library like `rustls`.
</details>

**Comment by briansmith** [comment]
> I/we couldn't think of a way to implement proper zero-copy scatter/gather without ring accepting a slice of mutable byte slices as contiguous memory for encryption (due to the issues mentioned earlier). At the same time I'm rather confident @briansmith won't want to support that because I assume it'd complicate ring's internals by a factor of `oh hello no`. _But_ I have a tendency to miss obvious things.

I am very open to extending the *ring* APIs to better support scatter/gather, zero-copy, and any other performance improvements. Internally *ring*'s AEAD stuff is implemented in terms of an I-U-F interface anyway,

As I mentioned elsewhere, I think we need to get agreement on what we want--what we think is optimal, and then we can schedule the work to get to that optimal state. Right now there's a lot of proposed incremental improvements but it isn't clear whether they are steps towards where we ultimately want to end up. I'd like to find a way to ensure as many of the incremental changes are moving towards the end we'd like.

**Comment by djc** [comment]
This was implemented as part of #710:

https://github.com/rustls/rustls/pull/710/commits/35d16f0c2b7f31ed95a4c4f67f7ac5db7bad776b

**Comment by cpu** [comment]
> This was implemented as part of https://github.com/rustls/rustls/pull/710:

I think based on #710 and #1050 this can be closed. If there's work left to be done it would probably benefit from an updated description after catching up with work that has landed in-tree since 2021.
